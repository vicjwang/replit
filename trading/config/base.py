from datetime import datetime


REGIME_START_DATE_DEFAULT = '2022-01-01'  # includes bear market
TICKER_REGIME_START_DATE = dict(
  MDB='2024-03-08',  # date of first < 200MA
)

IS_WIDESCREEN = True
IS_PHONE = False
MY_WIN_PROBA = 0.85
SKIP_GRAPHS = False
SHOW_GRAPHS = True
IS_DEBUG = True
IS_VERBOSE = False
SHOULD_AVOID_EARNINGS = True
MIN_EXPIRY_DATESTR = '2025-01-01'
USE_EARNINGS_CSV = False
MIN_ZSCORE_THRESHOLD = 0  #0.85  # 80% proba move

WORTHY_MIN_BID = 0.1  #0.8
WORTHY_MIN_ROI = 0  #0.2
MAX_STRIKE = 280

FROZEN_TEST_DATE = '2024-07-22'
CACHE_DIR = './cache'

TRADIER_THROTTLE_RATE = 119 # per minute
TRADIER_THROTTLE_PERIOD = 61  # in seconds

NOW = datetime.utcnow()
NUM_PARALLEL_JOBS = 4
FORCE_REFRESH = False

SIGNAL_MAX_PROBA_GLOBAL = 0.025
