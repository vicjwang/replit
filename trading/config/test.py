from datetime import datetime


CACHE_DIR = './tests/saved'
SHOW_GRAPHS = False
IS_DEBUG = True
MIN_ZSCORE = 0

#NOW = datetime(2024, 7, 22)
NOW = datetime(2024, 7, 23)
