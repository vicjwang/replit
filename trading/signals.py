

class Signal:
  def __init__(self, *args, weight=1, **kwargs):
    self.weight = weight


class SupportSignal(Signal):
  
  def __str__(self):
    raise NotImplementedError

  def validate_conditions(self, price, strike):
    return price > self.support_price and strike < self.support_price

  def compute_edge(self, row, max_proba, **kwargs):
    price = kwargs['price_model'].get_latest_price()
    strike = row['strike']
    if not self.validate_conditions(price, strike):
      return 0
    return self.weight * max_proba * (price - self.support_price) / (price - strike)


class MovingAverageSupportSignal(SupportSignal):
  def __init__(self, n, *args, **kwargs):
    super().__init__(*args, **kwargs)
    self.n = n

  def __str__(self):
    return "{}_ma_edge".format(self.n)

  def compute_edge(self, row, max_proba, **kwargs):
    self.support_price = kwargs['price_model'].get_ma(self.n)
    return super().compute_edge(price_model, *args)


class FiftyTwoLowSupportSignal(SupportSignal):
  def __str__(self):
    return '52_low_edge'

  def compute_edge(self, row, max_proba, **kwargs):
    self.support_price = kwargs['price_model'].get_52_low()
    return super().compute_edge(*args, **kwargs)


class DeltaSignal(Signal):
  
  def __str__(self):
    return 'delta_edge'
  
  def compute_edge(self, row, max_proba, **kwargs):
    lose_proba = 1 - kwargs['win_proba']
    delta = abs(row['delta'])
    print('vjw lose proba', lose_proba)
    print('vjw delta', delta)
    print('vjw row', row[['0.15_target', 'strike']])
    return max_proba * (1 - lose_proba / delta) if lose_proba < delta else 0
