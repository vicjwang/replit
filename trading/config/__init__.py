import os

from .base import *


ENV = os.environ.get('ENV')

if ENV == 'test':
  from .test import *
