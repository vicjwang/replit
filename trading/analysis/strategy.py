import matplotlib.dates as mdates
import numpy as np
import pandas as pd

import config

from analysis.models import PriceModel
from constants import (
  COVERED_CALLS,
  DATE_FORMAT,
  DELTA_UPPER,
  PHI_ZSCORE,
  MU,
  SIDE_SHORT,
  SIGMA_LOWER,
  MAX_STRIKE,
  WIN_PROBA_ZSCORE,
  WORTHY_MIN_BID,
  WORTHY_MIN_ROI,
)
from analysis.derivative_strategy import DerivativeStrategyBase


def sell_intraquarter_derivatives(symbol):
  if symbol in COVERED_CALLS:
    option_type = 'call'
  else:
    option_type = 'put'

  side = SIDE_SHORT

  deriv_strat = DerivativeStrategyBase(symbol, side=side)
  print(deriv_strat)
  price_model = deriv_strat.get_price_model()

  latest_price = price_model.get_latest_price()
  latest_change = price_model.get_latest_change()

  if (option_type == 'call' and latest_change < 0) or (option_type == 'put' and latest_change > 0):
    raise ValueError(f'{symbol} {option_type} move threshold not met. ${latest_price}, {round(latest_change * 100, 2)}%')

  next_earnings_date = price_model.get_next_earnings_date()

  return deriv_strat.build_snapshot(option_type, expiry_before=next_earnings_date)


def sell_LTDITM_puts(symbol):
  # Look at far away deep ITM Puts.
  side = SIDE_SHORT
  option_type = 'put'

  deriv_strat = DerivativeStrategyBase(symbol, side=side)
  print(deriv_strat)
  return deriv_strat.build_snapshot(option_type, expiry_after=config.MIN_EXPIRY_DATESTR)


def sell_LTDOTM_calls(symbol):
  # NOTE: YoY ROI generally not worth it (<.05)
  side = SIDE_SHORT
  option_type = 'call'

  deriv_strat = DerivativeStrategyBase(symbol, side=side)
  print(deriv_strat)
  return deriv_strat.build_snapshot(option_type, expiry_after=config.MIN_EXPIRY_DATESTR)
