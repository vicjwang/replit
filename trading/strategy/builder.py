import matplotlib.dates as mdates
import numpy as np
import pandas as pd

import config

from analysis.models import PriceModel
from constants import COVERED_CALLS, SIDE_SHORT
from strategy.credit_spreads import CreditSpreadStrategy
from utils import get_sig_level


class Build:
  
  def __init__(self, symbol, win_proba=config.MY_WIN_PROBA):
    self.symbol = symbol
    self.win_proba = win_proba
  
  def validate_conditions(self):
    pass

  def create_snapshot(self):
    pass


class SellSimpleCreditSpreadBuild(Build):

  def __init__(self, *args, **kwargs):
    super().__init__(*args, **kwargs)
    self.side = SIDE_SHORT
    self.strategy = CreditSpreadStrategy(self.symbol, side=self.side)
    self.price_model = self.strategy.get_price_model()
    self.option_type = 'put'

    self.validate_conditions()

  def validate_conditions(self):

    latest_price = self.price_model.get_latest_price()
    latest_change = self.price_model.get_latest_change()

    sigma = self.price_model.get_daily_stdev()
    min_change = sigma * config.MIN_ZSCORE_THRESHOLD

    if latest_change > min_change:
      raise ValueError(f'{self.symbol} {self.option_type} move threshold not met. ${latest_price}, {round(latest_change * 100, 2)}%')

  def create_snapshot(self):
    sig_level = get_sig_level(self.side, self.option_type, self.win_proba)
    next_earnings_date = self.price_model.get_next_earnings_date()
    return self.strategy.make_snapshot(self.option_type, sig_level, expiry_before=next_earnings_date)
  









def sell_intraquarter_derivatives(symbol, win_proba=config.MY_WIN_PROBA):
  if symbol in COVERED_CALLS:
    option_type = 'call'
  else:
    option_type = 'put'

  side = SIDE_SHORT

  deriv_strat = CreditSpreadStrategy(symbol, side=side)
  print(deriv_strat)
  price_model = deriv_strat.get_price_model()

  latest_price = price_model.get_latest_price()
  latest_change = price_model.get_latest_change()

  sigma = price_model.get_daily_stdev()
  min_change = sigma * config.MIN_ZSCORE_THRESHOLD

  if (option_type == 'call' and latest_change < min_change) or (option_type == 'put' and latest_change > -1*min_change):
    raise ValueError(f'{symbol} {option_type} move threshold not met. ${latest_price}, {round(latest_change * 100, 2)}%')

  next_earnings_date = price_model.get_next_earnings_date()

  sig_level = get_sig_level(side, option_type, win_proba)
  return deriv_strat.build_snapshot(option_type, sig_level, expiry_before=next_earnings_date)


def sell_LTDITM_puts(symbol, win_proba=config.MY_WIN_PROBA):
  # Look at far away deep ITM Puts.
  side = SIDE_SHORT
  option_type = 'put'

  deriv_strat = CreditSpreadStrategy(symbol, side=side)
  print(deriv_strat)
  return deriv_strat.build_snapshot(option_type, 0.15, expiry_after=config.MIN_EXPIRY_DATESTR)


def sell_LTDOTM_calls(symbol, win_proba=config.MY_WIN_PROBA):
  # NOTE: YoY ROI generally not worth it (<.05)
  side = SIDE_SHORT
  option_type = 'call'

  deriv_strat = CreditSpreadStrategy(symbol, side=side)
  print(deriv_strat)
  return deriv_strat.build_snapshot(option_type, 0.85, expiry_after=config.MIN_EXPIRY_DATESTR)
