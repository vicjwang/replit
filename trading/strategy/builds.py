import matplotlib.dates as mdates
import numpy as np
import pandas as pd

import config

from analysis.models import PriceModel
from constants import COVERED_CALLS, SIDE_SHORT
from strategy.base import DerivativeStrategyBase
from strategy.credit_spreads import CreditSpreadStrategy
from utils import get_sig_level


class Build:
  def __init__(self, symbol, win_proba, *args, signals=None, **kwargs):
    self.symbol = symbol
    self.win_proba = win_proba
    self._strategy = None
    self.signals = signals or []

  @property
  def strategy(self):
    raise NotImplementedError

  @property
  def price_model(self):
    return self.strategy.get_price_model()

  def validate_conditions(self):
    pass

  def create_snapshot(self):
    self.validate_conditions()
    snapshot = self._create_snapshot()

    signal_max_proba = (1 - self.win_proba) / len(self.signals)
    for signal in self.signals:
      snapshot.df[str(signal)] = snapshot.df.apply(lambda row: signal.compute_edge(row, signal_max_proba), axis=1)

    return snapshot

  def add_signals(self, signals):
    self.signals += signals


class SellSimplePutBuild(Build):
  def __init__(self, *args, **kwargs):
    super().__init__(*args, **kwargs)
    self.side = SIDE_SHORT
    self.option_type = 'put'

  @property
  def strategy(self):
    if self._strategy is None:
      self._strategy = DerivativeStrategyBase(self.symbol, side=self.side)
    return self._strategy

  def validate_conditions(self):
    latest_price = self.price_model.get_latest_price()
    latest_change = self.price_model.get_latest_change()

    sigma = self.price_model.get_daily_stdev()
    min_change = -1 * sigma * config.MIN_ZSCORE_THRESHOLD

    if latest_change > min_change:
      raise ValueError(f'{self.symbol} {self.option_type} move threshold not met. ${latest_price}, {round(latest_change * 100, 2)}%')

  def _create_snapshot(self):
    sig_level = get_sig_level(self.side, self.option_type, self.win_proba)
    next_earnings_date = self.price_model.get_next_earnings_date()
    return self.strategy.make_snapshot(self.option_type, sig_level, expiry_before=next_earnings_date)


class SellSimplePutCreditSpreadBuild(SellSimplePutBuild):

  @property
  def strategy(self):
    if self._strategy is None:
      self._strategy = CreditSpreadStrategy(self.symbol, side=self.side)
    return self._strategy


########## DEPRECATED ##########

def sell_LTDOTM_calls(symbol, win_proba=config.MY_WIN_PROBA):
  # NOTE: YoY ROI generally not worth it (<.05)
  side = SIDE_SHORT
  option_type = 'call'

  deriv_strat = CreditSpreadStrategy(symbol, side=side)
  print(deriv_strat)
  return deriv_strat.build_snapshot(option_type, 0.85, expiry_after=config.MIN_EXPIRY_DATESTR)
