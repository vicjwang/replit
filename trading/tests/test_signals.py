import pytest

import config

from datetime import datetime
from unittest.mock import patch

from strategy.builds import SellSimplePutBuild
from signals import (
  DeltaSignal,
  MovingAverageSupportSignal,
  FiftyTwoLowSupportSignal,
)


class TestSignalsComputeEdge:
  
  @pytest.fixture
  def build_ddog(self):
    """
    DDOG
    2024-10-23
    121.52, -2.49%
    52low = 77.81
    52high = 138.61
    50ma = 117.17
    200ma = 121.50
    """

    build = SellSimplePutBuild('DDOG', config.MY_WIN_PROBA)
    max_proba = 1 - config.MY_WIN_PROBA
    return build

  @pytest.fixture
  def build_nvda(self):
    """
    FIXME
    DDOG
    2024-10-23
    139.56, -2.81%
    52low = 39.22
    52high = 144.42
    50ma = 123.53
    200ma = 101.78
    """

    build = SellSimplePutBuild('NVDA', 0.85)
    max_proba = 1 - config.MY_WIN_PROBA
    return build

  @pytest.fixture
  def build_okta(self):
    """
    OKTA
    2024-10-23
    72.16, -2.53%
    52low = 65.74
    52high = 114.50
    50ma = 79.57 (price is below)
    200ma = 90.15 (price is below)
    """

    build = SellSimplePutBuild('OKTA', config.MY_WIN_PROBA)
    max_proba = 1 - config.MY_WIN_PROBA
    return build
  
  def test_200ma_nonzero(self, build_ddog, snapshot):
    build_ddog.add_signals([
      MovingAverageSupportSignal(200, weight=0.5)
    ])
    result_df = build_ddog.create_snapshot().df
    result = result_df.iloc[0]['200_ma_edge'].round(4)

    assert result_df.to_csv() == snapshot
    assert result == round(0.0001, 4)

  def test_200ma_zero(self, build_okta, snapshot):
    build_okta.add_signals([
      MovingAverageSupportSignal(200, weight=0.5)
    ])
    result_df = build_okta.create_snapshot().df
    result = result_df.iloc[0]['200_ma_edge'].round(4)

    assert result_df.to_csv() == snapshot
    assert result == 0

  def test_52_low_nonzero(self, build_okta, snapshot):
    build_okta.add_signals([
      FiftyTwoLowSupportSignal(weight=0.5)
    ])

    result_df = build_okta.create_snapshot().df
    result = result_df.iloc[0]['52_low_edge'].round(4)

    assert result_df.to_csv() == snapshot
    assert result == 0.0448

  def test_52_low_zero(self, build_ddog, snapshot):
    build_ddog.add_signals([
      FiftyTwoLowSupportSignal()
    ])
    result_df = build_ddog.create_snapshot().df
    result = result_df.iloc[0]['52_low_edge'].round(4)

    assert result_df.to_csv() == snapshot
    assert result == 0


  def test_delta_zero(self, build_ddog, snapshot):
    build_ddog.add_signals([
      DeltaSignal(),
    ])
    result_df = build_ddog.create_snapshot().df
    result = result_df.iloc[0]['delta_edge'].round(4)

    assert result_df.to_csv() == snapshot
    assert result == 0

  def test_delta_nonzero(self, build_nvda, snapshot):
    build_nvda.add_signals([
      DeltaSignal(),
    ])
    result_df = build_nvda.create_snapshot().df
    result1 = result_df.iloc[3]['delta_edge'].round(4)  # |delta| = 0.17
    result2 = result_df.iloc[1]['delta_edge'].round(4)  # |delta| = 0.14

    assert result_df.to_csv() == snapshot
    assert result1 == 0.0557
    assert result2 == 0

    
