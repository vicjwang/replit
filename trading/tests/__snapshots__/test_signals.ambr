# serializer version: 1
# name: TestSignalsComputeEdge.test_200ma_nonzero
  '''
  ,description,expiration_date,option_type,strike,bid,ask,volume,0.01_target,0.025_target,0.05_target,0.1_target,0.15_target,0.2_target,0.25_target,0.3_target,0.35_target,0.4_target,0.45_target,0.5_target,0.55_target,0.6_target,0.65_target,0.7_target,0.75_target,0.8_target,0.85_target,0.9_target,0.95_target,0.975_target,0.99_target,delta,gamma,theta,vega,rho,phi,bid_iv,mid_iv,ask_iv,smv_vol,updated_at,yoy_roi,200_ma_edge
  0,DDOG Nov 1 2024 $111.00 Put,2024-11-01,put,111.0,0.24,0.47,17,98.5120159056613,103.17150894508868,106.77162454342812,110.57486952695879,112.98724636429348,114.84194798825277,116.4018207425358,117.78627362122853,119.06149234095105,120.26974917127376,121.44154262793,122.60165793517365,123.77285567349466,124.9787800508983,126.24708655095012,127.61390666615227,129.13171316881824,130.88568063989018,133.03418759308323,135.93655224516132,140.77866280230268,145.6910602756951,152.58206209938615,-0.0900617667485879,0.019220490786684842,-0.09516700785225565,0.03246693253248297,0.02453964369428971,-0.02723828568917952,0.465219,0.47292,0.48062,0.478,2024-10-23 19:59:26,0.0876876876876877,0.0001425855513308039
  
  '''
# ---
# name: TestSignalsComputeEdge.test_200ma_zero
  '''
  ,description,expiration_date,option_type,strike,bid,ask,volume,0.01_target,0.025_target,0.05_target,0.1_target,0.15_target,0.2_target,0.25_target,0.3_target,0.35_target,0.4_target,0.45_target,0.5_target,0.55_target,0.6_target,0.65_target,0.7_target,0.75_target,0.8_target,0.85_target,0.9_target,0.95_target,0.975_target,0.99_target,delta,gamma,theta,vega,rho,phi,bid_iv,mid_iv,ask_iv,smv_vol,updated_at,yoy_roi,200_ma_edge
  0,OKTA Nov 8 2024 $65.00 Put,2024-11-08,put,65.0,0.25,0.34,0,57.832463333546556,60.31223677159187,62.345167318159504,64.60600641111081,66.09755808126849,67.27165731257855,68.27526214908302,69.17656044214579,70.01401234876377,70.8125704706785,71.59049892131742,72.36282770174469,73.14348847809408,73.94702378669655,74.79044061791791,75.69585419575337,76.695111350264,77.83930175321072,79.22197105306276,81.05095987006936,83.99013200606437,86.8211678638789,90.54393555383884,-0.1049012127138093,0.02824332890018419,-0.04625302078003723,0.028554807553528173,0.025042405603438858,-0.028429224565584832,0.390312,0.405475,0.420637,0.403,2024-10-23 20:00:06,0.08774038461538462,0
  1,OKTA Nov 15 2024 $64.00 Put,2024-11-15,put,64.0,0.34,0.4,1,56.27474247991504,58.863170715768675,61.046479594365984,63.53550215030525,65.20944710979876,66.54282737738409,67.69224098261103,68.73111503218054,69.70125210143311,70.63002601182954,71.53766775933745,72.44099007304725,73.35571884195758,74.29838751417726,75.28841856566065,76.35111172438145,77.52287362020374,78.86194875673205,80.47449066585907,82.59472051309065,85.96232047503104,89.15077082922991,93.25137373371867,-0.10966252324615,0.02467561445674862,-0.03906146789927802,0.033934986586454385,0.03503874647076872,-0.040611414853231054,0.393196,0.401418,0.409641,0.394,2024-10-23 20:00:06,0.08430706521739131,0
  2,OKTA Nov 22 2024 $63.00 Put,2024-11-22,put,63.0,0.37,1.35,0,54.867240669536095,57.58310153055943,59.91331456385473,62.61001327887444,64.44516079488145,65.9178556181614,67.19425065402893,68.35276084403785,69.43829998534905,70.48046842346093,71.50129800651878,72.51923687107114,73.55166777368234,74.61698019322385,75.73687313013338,76.9396824857185,78.26621571301345,79.78171721523019,81.6048816000506,83.99678327710885,87.7774790903534,91.32956677527248,95.85026788639833,-0.1298227739685653,0.021973661313406733,-0.04321583662413688,0.04265709955208253,0.04335989767840315,-0.05174533651342017,0.389054,0.488155,0.587255,0.47,2024-10-23 20:00:06,0.07145502645502645,0
  
  '''
# ---
# name: TestSignalsComputeEdge.test_52_low_nonzero
  '''
  ,description,expiration_date,option_type,strike,bid,ask,volume,0.01_target,0.025_target,0.05_target,0.1_target,0.15_target,0.2_target,0.25_target,0.3_target,0.35_target,0.4_target,0.45_target,0.5_target,0.55_target,0.6_target,0.65_target,0.7_target,0.75_target,0.8_target,0.85_target,0.9_target,0.95_target,0.975_target,0.99_target,delta,gamma,theta,vega,rho,phi,bid_iv,mid_iv,ask_iv,smv_vol,updated_at,yoy_roi,52_low_edge
  0,OKTA Nov 8 2024 $65.00 Put,2024-11-08,put,65.0,0.25,0.34,0,57.832463333546556,60.31223677159187,62.345167318159504,64.60600641111081,66.09755808126849,67.27165731257855,68.27526214908302,69.17656044214579,70.01401234876377,70.8125704706785,71.59049892131742,72.36282770174469,73.14348847809408,73.94702378669655,74.79044061791791,75.69585419575337,76.695111350264,77.83930175321072,79.22197105306276,81.05095987006936,83.99013200606437,86.8211678638789,90.54393555383884,-0.1049012127138093,0.02824332890018419,-0.04625302078003723,0.028554807553528173,0.025042405603438858,-0.028429224565584832,0.390312,0.405475,0.420637,0.403,2024-10-23 20:00:06,0.08774038461538462,0.044832402234636894
  1,OKTA Nov 15 2024 $64.00 Put,2024-11-15,put,64.0,0.34,0.4,1,56.27474247991504,58.863170715768675,61.046479594365984,63.53550215030525,65.20944710979876,66.54282737738409,67.69224098261103,68.73111503218054,69.70125210143311,70.63002601182954,71.53766775933745,72.44099007304725,73.35571884195758,74.29838751417726,75.28841856566065,76.35111172438145,77.52287362020374,78.86194875673205,80.47449066585907,82.59472051309065,85.96232047503104,89.15077082922991,93.25137373371867,-0.10966252324615,0.02467561445674862,-0.03906146789927802,0.033934986586454385,0.03503874647076872,-0.040611414853231054,0.393196,0.401418,0.409641,0.394,2024-10-23 20:00:06,0.08430706521739131,0.03933823529411767
  2,OKTA Nov 22 2024 $63.00 Put,2024-11-22,put,63.0,0.37,1.35,0,54.867240669536095,57.58310153055943,59.91331456385473,62.61001327887444,64.44516079488145,65.9178556181614,67.19425065402893,68.35276084403785,69.43829998534905,70.48046842346093,71.50129800651878,72.51923687107114,73.55166777368234,74.61698019322385,75.73687313013338,76.9396824857185,78.26621571301345,79.78171721523019,81.6048816000506,83.99678327710885,87.7774790903534,91.32956677527248,95.85026788639833,-0.1298227739685653,0.021973661313406733,-0.04321583662413688,0.04265709955208253,0.04335989767840315,-0.05174533651342017,0.389054,0.488155,0.587255,0.47,2024-10-23 20:00:06,0.07145502645502645,0.035043668122270756
  
  '''
# ---
# name: TestSignalsComputeEdge.test_52_low_zero
  '''
  ,description,expiration_date,option_type,strike,bid,ask,volume,0.01_target,0.025_target,0.05_target,0.1_target,0.15_target,0.2_target,0.25_target,0.3_target,0.35_target,0.4_target,0.45_target,0.5_target,0.55_target,0.6_target,0.65_target,0.7_target,0.75_target,0.8_target,0.85_target,0.9_target,0.95_target,0.975_target,0.99_target,delta,gamma,theta,vega,rho,phi,bid_iv,mid_iv,ask_iv,smv_vol,updated_at,yoy_roi,52_low_edge
  0,DDOG Nov 1 2024 $111.00 Put,2024-11-01,put,111.0,0.24,0.47,17,98.5120159056613,103.17150894508868,106.77162454342812,110.57486952695879,112.98724636429348,114.84194798825277,116.4018207425358,117.78627362122853,119.06149234095105,120.26974917127376,121.44154262793,122.60165793517365,123.77285567349466,124.9787800508983,126.24708655095012,127.61390666615227,129.13171316881824,130.88568063989018,133.03418759308323,135.93655224516132,140.77866280230268,145.6910602756951,152.58206209938615,-0.0900617667485879,0.019220490786684842,-0.09516700785225565,0.03246693253248297,0.02453964369428971,-0.02723828568917952,0.465219,0.47292,0.48062,0.478,2024-10-23 19:59:26,0.0876876876876877,0
  
  '''
# ---
# name: TestSignalsComputeEdge.test_delta_nonzero
  '''
  ,description,expiration_date,option_type,strike,bid,ask,volume,0.01_target,0.025_target,0.05_target,0.1_target,0.15_target,0.2_target,0.25_target,0.3_target,0.35_target,0.4_target,0.45_target,0.5_target,0.55_target,0.6_target,0.65_target,0.7_target,0.75_target,0.8_target,0.85_target,0.9_target,0.95_target,0.975_target,0.99_target,delta,gamma,theta,vega,rho,phi,bid_iv,mid_iv,ask_iv,smv_vol,updated_at,yoy_roi,delta_edge
  0,NVDA Oct 25 2024 $132.00 Put,2024-10-25,put,132.0,0.26,0.27,14772,99.71772635164169,113.92024343173497,122.0783521521754,128.56115820474037,131.81330292280145,133.97740923203327,135.6237997116937,136.98254646252818,138.1694171520034,139.2523603514337,140.2768814774473,141.27778867564297,142.28583757251954,143.3326768947234,144.45608865181663,145.70771305189288,147.1674854672185,148.97596309324254,151.42184537147352,155.25228499647793,163.49674795904397,175.20515205921257,200.1591322158243,-0.0730315031677871,0.02481079555438441,-0.24230561733601932,0.01370479392300849,0.006676454163371168,-0.0071125118950021715,0.580251,0.583413,0.586575,0.578,2024-10-23 19:59:53,0.359469696969697,0.0
  1,NVDA Nov 1 2024 $132.00 Put,2024-11-01,put,132.0,1.58,1.6,2402,112.86870776038724,118.86156619281753,123.51416590477146,128.44966647900094,131.59079832922777,134.01124895995153,136.0505587210443,137.86326986628205,139.5352109289949,141.1213332100238,142.66140798943783,144.18788391970938,145.73069313028802,147.3210704316594,148.99569600266017,150.8026459071267,152.81191098870337,155.1373189235524,157.99087879403695,161.854416902222,168.32195495108343,174.91058325378077,184.19760695213844,-0.2270630194763027,0.02629065347467151,-0.21486067023323804,0.0677844183533637,0.024314080934500323,-0.026654006539672537,0.520711,0.522923,0.525135,0.522,2024-10-23 19:59:53,0.48543771043771045,0.050908566917264114
  2,NVDA Nov 8 2024 $131.00 Put,2024-11-08,put,131.0,2.69,2.75,473,111.3116086000437,117.28726431447473,122.23145478990638,127.77656844803184,131.46114466081318,134.3759359702667,136.87739086666528,139.1315608716822,141.23252398875925,143.24166806449313,145.20426550647696,147.1579224457943,149.13786494513445,151.18124796488866,153.33192048798878,155.64731684807822,158.2106000227413,161.15574550010277,164.72893336229717,169.47907117547862,177.16760530899506,184.63602391218743,194.54802972413344,-0.2623560539953134,0.020207795160950136,-0.18256517633396102,0.0956822064031413,0.04013789507045742,-0.04532156111736185,0.546267,0.548356,0.550445,0.548,2024-10-23 19:59:53,0.4684398854961832,0.06423868571981979
  3,NVDA Nov 15 2024 $132.00 Put,2024-11-15,put,132.0,3.7,3.8,2135,109.6583859932271,115.9756785390929,121.35766025711058,127.55108375725925,131.750132539278,135.113905429212,138.02690195782426,140.67022915126512,143.14753710518843,145.52716826305732,147.86009138429722,150.18913898910665,152.55487304997516,155.00045620014527,157.57712585522097,160.35217690612774,163.4230512337492,166.9463805271101,171.20876492146564,176.84504753575197,185.8702402674869,194.4957576832438,205.7004329033475,-0.2967602683841574,0.01937083901211583,-0.1501962458026253,0.12098510476864015,0.0546163222057465,-0.062057769749657155,0.516774,0.520902,0.52503,0.52,2024-10-23 19:59:53,0.444828722002635,0.07418122505916576
  
  '''
# ---
# name: TestSignalsComputeEdge.test_delta_zero
  '''
  ,description,expiration_date,option_type,strike,bid,ask,volume,0.01_target,0.025_target,0.05_target,0.1_target,0.15_target,0.2_target,0.25_target,0.3_target,0.35_target,0.4_target,0.45_target,0.5_target,0.55_target,0.6_target,0.65_target,0.7_target,0.75_target,0.8_target,0.85_target,0.9_target,0.95_target,0.975_target,0.99_target,delta,gamma,theta,vega,rho,phi,bid_iv,mid_iv,ask_iv,smv_vol,updated_at,yoy_roi,delta_edge
  0,DDOG Nov 1 2024 $111.00 Put,2024-11-01,put,111.0,0.24,0.47,17,98.5120159056613,103.17150894508868,106.77162454342812,110.57486952695879,112.98724636429348,114.84194798825277,116.4018207425358,117.78627362122853,119.06149234095105,120.26974917127376,121.44154262793,122.60165793517365,123.77285567349466,124.9787800508983,126.24708655095012,127.61390666615227,129.13171316881824,130.88568063989018,133.03418759308323,135.93655224516132,140.77866280230268,145.6910602756951,152.58206209938615,-0.0900617667485879,0.019220490786684842,-0.09516700785225565,0.03246693253248297,0.02453964369428971,-0.02723828568917952,0.465219,0.47292,0.48062,0.478,2024-10-23 19:59:26,0.0876876876876877,0
  
  '''
# ---
